go/vt/tabletserver/grpctabletconn/conn.go

```diff
@@ -0,0 +1,307 @@
+func DialTablet(ctx context.Context, endPoint topo.EndPoint, keyspace, shard string, timeout time.Duration) (tabletconn.TabletConn, error) {
+	// create the RPC client
+	addr := netutil.JoinHostPort(endPoint.Host, endPoint.NamedPortMap["grpc"])
+	cc, err := grpc.Dial(addr)
+	if err != nil {
+		return nil, err
+	}
+	c := pbs.NewQueryClient(cc)
+
+	gsir, err := c.GetSessionId(ctx, &pb.GetSessionIdRequest{
+		Keyspace: keyspace,
+		Shard:    shard,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if gsir.Error != nil {
+		return nil, tabletErrorFromRPCError(gsir.Error)
+	}
+
+	return &gRPCQueryClient{
+		endPoint:  endPoint,
+		cc:        cc,
+		c:         c,
+		sessionID: gsir.SessionId,
+	}, nil
+}
+
+// Execute sends the query to VTTablet.
+func (conn *gRPCQueryClient) Execute(ctx context.Context, query string, bindVars map[string]interface{}, transactionID int64) (*mproto.QueryResult, error) {
+	conn.mu.RLock()
+	defer conn.mu.RUnlock()
+	if conn.cc == nil {
+		return nil, tabletconn.ConnClosed
+	}
+
+	req := &pb.ExecuteRequest{
+		Query:         tproto.BoundQueryToProto3(query, bindVars),
+		TransactionId: transactionID,
+		SessionId:     conn.sessionID,
+	}
+	er, err := conn.c.Execute(ctx, req)
+	if err != nil {
+		return nil, tabletErrorFromGRPC(err)
+	}
+	if er.Error != nil {
+		return nil, tabletErrorFromRPCError(er.Error)
+	}
+	return tproto.Proto3ToQueryResult(er.Result), nil
+}
+
+// ExecuteBatch sends a batch query to VTTablet.
+func (conn *gRPCQueryClient) ExecuteBatch(ctx context.Context, queries []tproto.BoundQuery, transactionID int64) (*tproto.QueryResultList, error) {
+	conn.mu.RLock()
+	defer conn.mu.RUnlock()
+	if conn.cc == nil {
+		return nil, tabletconn.ConnClosed
+	}
+
+	req := &pb.ExecuteBatchRequest{
+		Queries:       make([]*pb.BoundQuery, len(queries)),
+		TransactionId: transactionID,
+		SessionId:     conn.sessionID,
+	}
+	for i, q := range queries {
+		req.Queries[i] = tproto.BoundQueryToProto3(q.Sql, q.BindVariables)
+	}
+	ebr, err := conn.c.ExecuteBatch(ctx, req)
+	if err != nil {
+		return nil, tabletErrorFromGRPC(err)
+	}
+	if ebr.Error != nil {
+		return nil, tabletErrorFromRPCError(ebr.Error)
+	}
+	return tproto.Proto3ToQueryResultList(ebr.Results), nil
+}
+
+// StreamExecute starts a streaming query to VTTablet.
+func (conn *gRPCQueryClient) StreamExecute(ctx context.Context, query string, bindVars map[string]interface{}, transactionID int64) (<-chan *mproto.QueryResult, tabletconn.ErrFunc, error) {
+	conn.mu.RLock()
+	defer conn.mu.RUnlock()
+	if conn.cc == nil {
+		return nil, nil, tabletconn.ConnClosed
+	}
+
+	req := &pb.StreamExecuteRequest{
+		Query:     tproto.BoundQueryToProto3(query, bindVars),
+		SessionId: conn.sessionID,
+	}
+	stream, err := conn.c.StreamExecute(ctx, req)
+	if err != nil {
+		return nil, nil, tabletErrorFromGRPC(err)
+	}
+	sr := make(chan *mproto.QueryResult, 10)
+	var finalError error
+	go func() {
+		for {
+			ser, err := stream.Recv()
```

Comment:
Here's where you were running into the issues with panic/errfunc. The other implementation check the first message before going into the loop, and `return nil, nil, tabletError(err)` if the first message was not received successfully. You're treating the first message the same as all the others.

I believe the first message is considered special because that's the one with the Fields, although I don't think it should actually matter.